### PR TITLE
clang-tidy only relevant sources when headers change

### PR DESCRIPTION
**Pull request by akrieger**
#### Summary
None

#### Purpose of change
The `clang-tidy` checks as currently implemented take forever if a header is changed. This is because we didn't have the ability to precisely compute which sources depended on that header. So, on any diffs changing a header, clang-tidy would run over the entire codebase. Kinda slow. If we could figure out exactly sources need to be linted on any change, we could more plausibly rely on `clang-tidy` to give speedy signal at PR time.

#### Describe the solution
clang and gcc support a `-H` flag, which makes them print to `stdout` a textual visualization of the include graph for a given source file. We extend the root Makefile to support `make includes`, which creates a set of `.inc` files in the output folder consisting of this output. Then a python helper script `get_affected_files.py` takes as input a filename consisting of 'affected files' and prints which sources include any of those affected files. Then we wire that up to the clang-tidy infra in `build-scripts/build.sh` and `build-scripts/clang-tidy.sh` instead of the fallback logic to 'lint everything on a header change'.

#### Testing
Ran it locally with some inputs and the results are sensible.
```
akrieger@SCOTTY:~/Cataclysm-DDA$ echo "src/main_menu.h" > files_changed
akrieger@SCOTTY:~/Cataclysm-DDA$ build-scripts/get_affected_files.py files_changed
src/main_menu.cpp
src/main.cpp

akrieger@SCOTTY:~/Cataclysm-DDA$ echo "src/mondeath.h" > files_changed
akrieger@SCOTTY:~/Cataclysm-DDA$ build-scripts/get_affected_files.py files_changed
src/mondeath.cpp
src/monstergenerator.cpp
src/monster.cpp
src/mtype.cpp

akrieger@SCOTTY:~/Cataclysm-DDA$ echo "src/mondeath.h" > files_changed
akrieger@SCOTTY:~/Cataclysm-DDA$ CATA_CLANG_TIDY=plugin COMPILER=clang++-12 TILES=1 SOUND=1 build-scripts/clang-tidy.sh
...
Analyzing affected files
6185 warnings generated.
7289 warnings generated.
7985 warnings generated.
7059 warnings generated
```

#### Additional context

Currently assuming only clang/gcc (msvc has a similar mechanism) and all sources end in .cpp. These are all also assumptions the Makefile already has so that doesn't feel like a bad thing to perpetuate.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -466,8 +466,9 @@ ifeq ($(PCH), 1)
   endif
 endif
 
+DEPFLAGS=-MMD -MP
```
Comment (BrettDong) [inline Makefile:469]: Variable assignments in the rest of Makefile usually have spaces.
```suggestion
DEPFLAGS = -MMD -MP
```

File: build-scripts/get_files_including.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    lintable_files = set()
+    with open(changed_files_file, 'r') as f:
+        changed_files = f.readlines()
+        for changed_file in changed_files:
+            for affected_file in includes_to_sources.get(changed_file, []):
+                lintable_files.add(affected_file)
+
+    for lintable_file in lintable_files:
+        print(lintable_file)
+
```
Comment (BrettDong) [inline build-scripts/get_files_including.py:43]: Python LGTM.com check suggests adding a return value
```suggestion

    return 0

```

**Comment by jbytheway** [comment]
I think clang-tidy runs on the tiles build, so this should probably be checking the tiles build too (or at least have that option).  If it's just in the Makefile, can it easily support all the same build options the regular build does?

**Comment by akrieger** [comment]
You're right about it being a TILES & SOUND build, that's new since I last ran clang-tidy. I'll update the script's assumptions for that.

The depfile recipe reuses all the same flags that compilation would use, it just swaps out the Make-depfile creation flags for the header trace & preprocess-only flags.

File: build-scripts/build.sh [hunk 3]
```diff
@@ -138,23 +138,46 @@ then
...
-        if [ -n "$changed_cpp_files" ]
+        if [ "$TIDY" == "all" ]
         then
-            remaining_cpp_files="$( \
-                echo "$all_cpp_files" | grep -v -F "$changed_cpp_files" || true )"
+            echo "Analyzing all files"
+            tidyable_cpp_files=$all_cpp_files
         else
-            remaining_cpp_files="$all_cpp_files"
+            if [ -f ./files_changed ]
```
Comment (akrieger) [inline build-scripts/build.sh:160]: Whups.

**Comment by akrieger** [comment]
Whups my technically correct change was masked by the fact that I changed the actual `clang-tidy.sh` script which then made it run over everything anyway. Derp.

**Comment by akrieger** [comment]
This misses test files which the original code covers. Should be easy to fix, but not ready yet.

**Comment by akrieger** [comment]
Now also handles test srcs as well as regular srcs.

**Comment by jbytheway** [comment]
> Now also handles test srcs as well as regular srcs.

Does it also cover the utilities (like the json formatter code)?

**Comment by akrieger** [comment]
I thought to myself 'surely if there is one generic recipe for .o files, every .o file goes through that recipe'.

The existing clang-tidy checks only cover cata src, tests, and the clang-tidy plugin itself. I could extend coverage to 'almost everything the makefile builds' and it'll probably encourage me to refactor the makefile for better DRY too.

**Comment by akrieger** [comment]
Actually, I can't do that unless I extend the cmake build to be able to handle it. clang-tidy depends on the compiledb from cmake to be able to execute properly.

This feels a bit too scope-creepy for me. If any of the clang-tidy sources change the infra already falls back to rerunning tidy over everything. Otherwise it'll properly scope down to tests and srcs like it already does.

**Comment by BrettDong** [comment]
If I remember correctly, the compile db used by `clang-tidy` check is faked from `Makefile`.

https://github.com/CleverRaven/Cataclysm-DDA/blob/4b90b94ae4cf77d3ad33bc53455d7ad9d6afd387/build-scripts/clang-tidy.sh#L77

**Comment by akrieger** [comment]
The `compiledb` invocation is executed within the `build` directory that cmake runs in and operates over the Makefile that cmake `generates` earlier in the script.

I made my prior statement after dumping `all_cpp_files` from [here](https://github.com/CleverRaven/Cataclysm-DDA/blob/4b90b94ae4cf77d3ad33bc53455d7ad9d6afd387/build-scripts/clang-tidy.sh#L85) and verifying what sources are actually listed.

**Comment by akrieger** [comment]
And, honestly, it's not clear we even need the compiledb call anymore. We invoke cmake asking it to create a `compile_commands.json` file regardless. I don't see why compiledb's generated output, based on the generated Makefile that cmake make anyway,

https://github.com/CleverRaven/Cataclysm-DDA/blob/4b90b94ae4cf77d3ad33bc53455d7ad9d6afd387/build-scripts/clang-tidy.sh#L44

**Comment by jbytheway** [comment]
The problem with building the compilation database via CMake directly (rather than via compiledb) is that CMake will actually build the code, which takes a long time, whereas compiledb fakes it and builds the database almost instantly.

It would be good if the clang-tidy checks continued to cover the clang-tidy plugin itself, but the dependencies for those can't be captured through the regular Makefile (and I don't think it's reasonable to try to capture them there, because building the plugin would be hard to do in raw make).  This is why I was originally hoping we could capture the dependencies via CMake, because then it would be easy to be consistent.

We could hack something together for the plugin sources, since there's only a couple of headers there and generally all the sources include all the headers.

**Comment by jbytheway** [comment]
> Doesn't CMake generate the compile database as long as the project is configured with -DCMAKE_EXPORT_COMPILE_COMMANDS=ON?

Yes, but it exports them at compile time, not configure time.  At least, that's what I thought I had observed, but I haven't tested this in years...

**Comment by akrieger** [comment]
> > Doesn't CMake generate the compile database as long as the project is configured with -DCMAKE_EXPORT_COMPILE_COMMANDS=ON?
> 
> 
> 
> Yes, but it exports them at compile time, not configure time.  At least, that's what I thought I had observed, but I haven't tested this in years...

From what I'm seeing here, it does it at makefile generation time, so it happens during the clang-tidy script invocation we already have.

I guess I can try to look and see what covering the extra bits will take. I don't have too much spare motivation lately so that might take a while. What's here already works.

**Comment by BrettDong** [comment]
If I remember correctly, `cmake .. -DCMAKE_EXPORT_COMPILE_COMMANDS=ON` generates the `compile_commands.json` during CMake configuration time.

**Comment by akrieger** [comment]
[#digressions] The cmake docs for that flag say that it is only support for Make/Ninja generators ("Note This option is implemented only by [Makefile Generators](https://cmake.org/cmake/help/latest/manual/cmake-generators.7.html#makefile-generators) and the [Ninja](https://cmake.org/cmake/help/latest/generator/Ninja.html#generator:Ninja). It is ignored on other generators.")  so at least some part of the logic is dependent on the actual generator backend and not the abstract configure logic. Unless that's what you mean by 'configuration time' because I don't know cmake well enough.

**Comment by akrieger** [comment]
Sorry, I ended up getting myself mixed up and missed a key question.

> It would be good if the clang-tidy checks continued to cover the clang-tidy plugin itself

It will. I preserved the logic which falls back to 'lint all files if certain changes are detected', which includes any changes to files containing `clang-tidy`, which includes sources under `tools/clang-tidy-plugin`. All I did was remove the wildcard `*.h` from that regexp. The original regexp is here.

https://github.com/CleverRaven/Cataclysm-DDA/blob/4b90b94ae4cf77d3ad33bc53455d7ad9d6afd387/build-scripts/clang-tidy.sh#L114-L116

The root CMakeLists.txt includes an option for JSON_FORMATTER, and chkjson seems to be implicitly included when not on msvc, so I'll extend the makefile depfile calculations to include those as well and see about explicitly adding JSON_FORMATTER to the clang-tidy checks.

I don't think there's a way for us to get cmake to emit the include tree / depfiles information like I coded into the Makefile, so we're stuck with doing that. The cmake generated Makefile doesn't even properly do depfile generation for Make ingestion, afaict.

**Comment by jbytheway** [comment]
> If I remember correctly, `cmake .. -DCMAKE_EXPORT_COMPILE_COMMANDS=ON` generates the `compile_commands.json` during CMake configuration time.

I finally had the time to go test this myself, and you are correct.  I was wrong.  The `compile_commands.json` is generated when running `cmake` (i.e. what I was calling "configure time"), which makes our lives easier.

**Pull request merged**